plugins/GranularPitchShifter/GranularPitchShifterEffect.cpp

```diff
@@ -0,0 +1,269 @@
+{
+	autoQuitModel()->setValue(autoQuitModel()->maxValue());
+	
+	changeSampleRate();
+}
+
+
+bool GranularPitchShifterEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
+{
+	if (!isEnabled() || !isRunning()) { return false; }
+
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const ValueBuffer * pitchBuf = m_granularpitchshifterControls.m_pitchModel.valueBuffer();
+	const ValueBuffer * pitchSpreadBuf = m_granularpitchshifterControls.m_pitchSpreadModel.valueBuffer();
+
+	const float size = m_granularpitchshifterControls.m_sizeModel.value();
+	const float shape = m_granularpitchshifterControls.m_shapeModel.value();
+	const float jitter = m_granularpitchshifterControls.m_jitterModel.value();
+	const float twitch = m_granularpitchshifterControls.m_twitchModel.value();
+	const float spray = m_granularpitchshifterControls.m_sprayModel.value();
+	const float spraySpread = m_granularpitchshifterControls.m_spraySpreadModel.value();
+	const float density = m_granularpitchshifterControls.m_densityModel.value();
+	const float glide = m_granularpitchshifterControls.m_glideModel.value();
+	const int minLatency = m_granularpitchshifterControls.m_minLatencyModel.value() * m_sampleRate;
+	const float densityInvRoot = std::sqrt(1.f / density);
+	const float feedback = m_granularpitchshifterControls.m_feedbackModel.value() * densityInvRoot;
+	const float fadeLength = 1.f / m_granularpitchshifterControls.m_fadeLengthModel.value();
+	const bool prefilter = m_granularpitchshifterControls.m_prefilterModel.value();
+	
+	if (glide != m_oldGlide)
+	{
+		m_oldGlide = glide;
+		m_glideCoef = std::exp(-1 / (glide * m_sampleRate));
+	}
+	
+	const float shapeK = cosWindowApproxK(shape);
+	const int sizeSamples = m_sampleRate / size;
+	const float waitMult = sizeSamples / (density * 2);
+
+	for (fpp_t f = 0; f < frames; ++f)
+	{
+		const double pitch = pitchBuf ? pitchBuf->value(f) : m_granularpitchshifterControls.m_pitchModel.value();
+		const double pitchSpread = (pitchSpreadBuf ? pitchSpreadBuf->value(f) : m_granularpitchshifterControls.m_pitchSpreadModel.value()) * 0.5f;
+		
+		// interpolate pitch depending on glide
+		for (int i = 0; i < 2; ++i)
+		{
+			double targetVal = pitch + pitchSpread * (i ? 1. : -1.);
+			
+			if (targetVal == m_truePitch[i]) { continue; }
+			m_updatePitches = true;
+			
+			m_truePitch[i] = m_glideCoef * m_truePitch[i] + (1. - m_glideCoef) * targetVal;
+			// we crudely lock the pitch to the target value once it gets close enough, so we can save on CPU
+			if (std::abs(targetVal - m_truePitch[i]) < GlideSnagRadius) { m_truePitch[i] = targetVal; }
+		}
+		
+		// this stuff is computationally expensive, so we should only do it when necessary
+		if (m_updatePitches)
+		{
+			m_updatePitches = false;
+			
+			std::array<double, 2> speed = {std::exp2(m_truePitch[0] * (1. / 12.)), std::exp2(m_truePitch[1] * (1. / 12.))};
+			std::array<double, 2> ratio = {speed[0] / m_speed[0], speed[1] / m_speed[1]};
```

Comment:
A little easier to read
```suggestion
			std::array<double, 2> speed = {
				std::exp2(m_truePitch[0] * (1. / 12.)),
				std::exp2(m_truePitch[1] * (1. / 12.))
			};
			std::array<double, 2> ratio = {
				speed[0] / m_speed[0],
				speed[1] / m_speed[1]
			};
```